### ISSUE TITLE
[css-animations] Animations should be allowed on linear/radial gradients

### ISSUE THREAD

**Issue by SetTrend** [w3c/csswg-drafts#3783]
CSS animations should explicitly allow animated `linear-gradient()` and `radial-gradient()` backgrounds.

If all `<keyframe-selector>`s of an animation yield computed background values sharing the same set of properties:
- either `linear-gradient()`, `radial-gradient`, `repeating-linear-gradient()` or `repeating-radial-gradient()`
- same number of `<color-stop>` values

Then animation should take place in the form that the `<color-stop>` values should get animated over time.

Internet Explorer and Microsoft Edge already implement this behaviour:

![Throbbers](https://user-images.githubusercontent.com/9283914/55298060-b647da80-542b-11e9-86c1-2b71086b7576.gif)

### Sample Code

```html
<!DOCTYPE html>
<html lang="en">

  <head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta http-equiv="X-UA-Compatible" content="ie=edge">
    <title>Document</title>
    <style>
      div {
        margin: 1em;
      }

      @keyframes movement1 {
        from {
          background: linear-gradient(to right, white, white 0%, #aaf 0%, white 2%, white);
        }

        25% {
          background: linear-gradient(to right, white, white 25%, #aaf 50%, white 52%, white);
        }

        50% {
          background: linear-gradient(to right, white, white 98%, #aaf 99%, white 100%, white);
        }

        75% {
          background: linear-gradient(to right, white, white 48%, #aaf 50%, white 75%, white);
        }

        to {
          background: linear-gradient(to right, white, white 0%, #aaf 1%, white 2%, white);
        }
      }

      @keyframes movement2 {
        from {
          background: radial-gradient(circle closest-side at center, #aaf, white, white);
        }

        50% {
          background: radial-gradient(circle closest-side at center, white, #aaf, white);
        }

        to {
          background: radial-gradient(circle closest-side at center, white, white, white);
        }
      }

      input[type="text"]#i1 {
        animation: 2s linear alternate infinite movement1;
        border: 1px solid silver;
        border-radius: .5ex;
      }

      input[type="text"]#i2 {
        animation: 2s linear alternate infinite movement2;
        border: 1px solid silver;
        border-radius: .5ex;
      }
    </style>
  </head>

  <body>
    <div>
      <input type="text" id="i1" />
    </div>
    <div>
      <input type="text" id="i2" />
    </div>
  </body>

</html>
```

**Comment by AmeliaBR** [comment]
Hi @SetTrend.  I think everyone in the working group agrees with you that this _should_ be possible. The problem is that there are a lot of little details that still need to be sorted out.

The relevant specification is [CSS Images Level 4](https://drafts.csswg.org/css-images-4/#interpolating-gradients).  It was initially described in Level 3, but got pushed back because browser implementations weren't always consistent. (For a while, Chrome also supported some gradient animations.)  Also, the current definition relies on falling back to `cross-fade()` animations if the gradients don't match up nicely, and no browsers yet support `cross-fade()`. 

So, what can you do?

- Take a look at the current spec text & file issues if you have any suggestions or concerns.
- File or upvote issues on the browser engines asking them to implement experimental versions of gradient interpolation as defined in CSS Images 4, so that we can get feedback and figure out the missing spec details.

**Comment by SetTrend** [comment]
Hi @AmeliaBR ,

thanks for replying.

I couldn't find corresponding issues yet, so I created two issues now, both back-referencing this issue:

- https://bugs.chromium.org/p/chromium/issues/detail?id=948090
- https://bugzilla.mozilla.org/show_bug.cgi?id=1540583

Anyone who's interested, please upvote there and here.

**Comment by SetTrend** [comment]
Just let me add an example depicting that IE and MS Edge even interpolate angle and color of linear gradients:

![Throbber](https://user-images.githubusercontent.com/9283914/55364358-a5a56c00-54e0-11e9-8fa0-38868d0e65c0.gif)

```html
<!DOCTYPE html>
<html lang="en">

	<head>
		<meta charset="UTF-8">
		<meta name="viewport" content="width=device-width, initial-scale=1.0">
		<meta http-equiv="X-UA-Compatible" content="ie=edge">
		<title>Document</title>
		<style>
			div
			{
				margin: 1em;
			}

			@keyframes movement1
			{
				from
				{
					background: linear-gradient(0deg, white, white 0%, #aaf 0%, white 2%, white);
				}

				25%
				{
					background: linear-gradient(90deg, white, white 25%, red 50%, white 52%, white);
				}

				50%
				{
					background: linear-gradient(180deg, white, white 98%, #aaf 99%, white 100%, white);
				}

				75%
				{
					background: linear-gradient(90deg, white, white 48%, purple 50%, white 75%, white);
				}

				to
				{
					background: linear-gradient(0deg, white, white 0%, #aaf 1%, white 2%, white);
				}
			}
			input[type="text"]#i1
			{
				animation: 2s linear infinite movement1;
				border: 1px solid silver;
				border-radius: .5ex;
			}
		</style>
	</head>

	<body>
		<div>
			<input type="text" id="i1" />
		</div>
	</body>

</html>
```

**Comment by stephenmcgruer** [comment]
As a drive-by note, the css-background spec actually doesn't allow interpolation of background-images - it is [defined as discretely interpolated](https://drafts.csswg.org/css-backgrounds-3/#background-image).

See issue 578 where alancutter long ago tried to get some attention to this.

**Comment by SetTrend** [comment]
@stephenmcgruer:

This issue is in fact about amending the spec to allow animations on gradients, which can be interpolated fine.

I couldn't think of any kinf of animation on background-images (i.e. discrete image files) other than fading.

**Comment by stephenmcgruer** [comment]
Right, but a `<gradient>` is an image type (unless I'm badly mistaken!). As such, in the examples which use `background: linear-gradient(...)` are actually equivalent to using `background-image: linear-gradient(...)`.

What I was pointing out is that now matter how we spec interpolation of gradients, unless we fix https://github.com/w3c/csswg-drafts/issues/578 we will not be able to interpolate the `background-image` property (and thus gradients in `background-image`) by spec.

**Comment by AmeliaBR** [comment]
Under the new definition of "animatable", background-image is [animatable, but as a discrete value](https://drafts.csswg.org/css-backgrounds-3/#background-image).  If we introduce a way for certain types of images to animate continuously, yes, that would need to be updated.

**Comment by SetTrend** [comment]
Wouldn't this (I'm quoting your phrase now) "certain type of image" be a _vector_ image in contrast to _raster_ images?

Probably only a certain fraction of vector images may be able to be faded - vector images sharing the same number of primitives, for instance.

There are algorithms available that morph different primitive shapes into each other. A circle, for instance, can be morphed into a hexagon. I guess the same applies to fill/pen brushes, too (as we can see from the Internet Explorer rendering above).
